Summit Compiler Error

Just tried comping Cabana on Summit today. Getting the following compiler error:

```
[ 32%] Building CXX object core/unit_test/CMakeFiles/Parallel_test_SERIAL.dir/unit_test_main.cpp.o
/ccs/home/uy7/software/Cabana/core/src/Cabana_Parallel.hpp: In instantiation of 'Cabana::neighbor_parallel_for(const Kokkos::RangePolicy<Traits ...>&, const FunctorType&, const NeighborListType&, Cabana::SecondNeighborsTag, Cabana::TeamOpTag, const string&)::<lambda(const typename Kokkos::TeamPolicy<typename Kokkos::RangePolicy<Traits ...>::execution_space, Kokkos::Schedule<Kokkos::Dynamic> >::member_type&)>::<lambda(index_type)> [with FunctorType = __nv_hdl_wrapper_t<false, false, __nv_dl_tag<void (*)(), Test::testAngularParallelFor<Cabana::VerletLayoutCSR>, 2u>, void(int, int, int), Kokkos::View<int*, Kokkos::CudaSpace> >; NeighborListType = Cabana::VerletList<Kokkos::CudaSpace, Cabana::FullNeighborTag, Cabana::VerletLayoutCSR>; ExecParameters = {Kokkos::Cuda}; index_type = unsigned int]':
/ccs/home/uy7/software/Cabana/core/src/Cabana_Parallel.hpp:423:89:   required from 'struct Cabana::neighbor_parallel_for(const Kokkos::RangePolicy<Traits ...>&, const FunctorType&, const NeighborListType&, Cabana::SecondNeighborsTag, Cabana::TeamOpTag, const string&)::<lambda(const typename Kokkos::TeamPolicy<typename Kokkos::RangePolicy<Traits ...>::execution_space, Kokkos::Schedule<Kokkos::Dynamic> >::member_type&)> [with FunctorType = __nv_hdl_wrapper_t<false, false, __nv_dl_tag<void (*)(), Test::testAngularParallelFor<Cabana::VerletLayoutCSR>, 2u>, void(int, int, int), Kokkos::View<int*, Kokkos::CudaSpace> >; NeighborListType = Cabana::VerletList<Kokkos::CudaSpace, Cabana::FullNeighborTag, Cabana::VerletLayoutCSR>; ExecParameters = {Kokkos::Cuda}; typename Kokkos::TeamPolicy<typename Kokkos::RangePolicy<Traits ...>::execution_space, Kokkos::Schedule<Kokkos::Dynamic> >::member_type = Kokkos::Impl::CudaTeamMember]::<lambda(using index_type = unsigned int)>'
/ccs/home/uy7/software/Cabana/core/src/Cabana_Parallel.hpp:412:21:   required from 'Cabana::neighbor_parallel_for(const Kokkos::RangePolicy<Traits ...>&, const FunctorType&, const NeighborListType&, Cabana::SecondNeighborsTag, Cabana::TeamOpTag, const string&)::<lambda(const typename Kokkos::TeamPolicy<typename Kokkos::RangePolicy<Traits ...>::execution_space, Kokkos::Schedule<Kokkos::Dynamic> >::member_type&)> [with FunctorType = __nv_hdl_wrapper_t<false, false, __nv_dl_tag<void (*)(), Test::testAngularParallelFor<Cabana::VerletLayoutCSR>, 2u>, void(int, int, int), Kokkos::View<int*, Kokkos::CudaSpace> >; NeighborListType = Cabana::VerletList<Kokkos::CudaSpace, Cabana::FullNeighborTag, Cabana::VerletLayoutCSR>; ExecParameters = {Kokkos::Cuda}; typename Kokkos::TeamPolicy<typename Kokkos::RangePolicy<Traits ...>::execution_space, Kokkos::Schedule<Kokkos::Dynamic> >::member_type = Kokkos::Impl::CudaTeamMember]'
/ccs/home/uy7/software/Cabana/core/src/Cabana_Parallel.hpp:423:89:   required from 'struct Cabana::neighbor_parallel_for(const Kokkos::RangePolicy<Traits ...>&, const FunctorType&, const NeighborListType&, Cabana::SecondNeighborsTag, Cabana::TeamOpTag, const string&) [with FunctorType = __nv_hdl_wrapper_t<false, false, __nv_dl_tag<void (*)(), Test::testAngularParallelFor<Cabana::VerletLayoutCSR>, 2u>, void(int, int, int), Kokkos::View<int*, Kokkos::CudaSpace> >; NeighborListType = Cabana::VerletList<Kokkos::CudaSpace, Cabana::FullNeighborTag, Cabana::VerletLayoutCSR>; ExecParameters = {Kokkos::Cuda}; std::__cxx11::string = std::__cxx11::basic_string<char>]::<lambda(const member_type&)>'
/ccs/home/uy7/software/Cabana/core/src/Cabana_Parallel.hpp:409:66:   required from 'void Cabana::neighbor_parallel_for(const Kokkos::RangePolicy<Traits ...>&, const FunctorType&, const NeighborListType&, Cabana::SecondNeighborsTag, Cabana::TeamOpTag, const string&) [with FunctorType = __nv_hdl_wrapper_t<false, false, __nv_dl_tag<void (*)(), Test::testAngularParallelFor<Cabana::VerletLayoutCSR>, 2u>, void(int, int, int), Kokkos::View<int*, Kokkos::CudaSpace> >; NeighborListType = Cabana::VerletList<Kokkos::CudaSpace, Cabana::FullNeighborTag, Cabana::VerletLayoutCSR>; ExecParameters = {Kokkos::Cuda}; std::__cxx11::string = std::__cxx11::basic_string<char>]'
/tmp/tmpxft_000178fb_00000000-5_tstNeighborList_CUDA.cudafe1.stub.c:80:74:   required from here
/ccs/home/uy7/software/Cabana/core/src/Cabana_Parallel.hpp:418:23: internal compiler error: in maybe_undo_parenthesized_ref, at cp/semantics.c:1694
                     for ( int a = n + 1; a < nn; ++a )
                     ~~^~~~~~
0x102f7e83 maybe_undo_parenthesized_ref(tree_node*)
	/sw/summit/gcc/6.4.0/src/gcc-6.4.0/gcc/cp/semantics.c:1693
0x10354947 cp_fold
	/sw/summit/gcc/6.4.0/src/gcc-6.4.0/gcc/cp/cp-gimplify.c:2069
0x102cc3e7 cp_build_binary_op(unsigned int, tree_code, tree_node*, tree_node*, int)
	/sw/summit/gcc/6.4.0/src/gcc-6.4.0/gcc/cp/typeck.c:5104
0x101de3af build_new_op_1
	/sw/summit/gcc/6.4.0/src/gcc-6.4.0/gcc/cp/call.c:5791
0x101dee5f build_new_op(unsigned int, tree_code, int, tree_node*, tree_node*, tree_node*, tree_node**, int)
	/sw/summit/gcc/6.4.0/src/gcc-6.4.0/gcc/cp/call.c:5836
0x102bfb37 build_x_binary_op(unsigned int, tree_code, tree_node*, tree_code, tree_node*, tree_code, tree_node**, int)
	/sw/summit/gcc/6.4.0/src/gcc-6.4.0/gcc/cp/typeck.c:3866
0x1021fa37 tsubst_copy_and_build(tree_node*, tree_node*, int, tree_node*, bool, bool)
	/sw/summit/gcc/6.4.0/src/gcc-6.4.0/gcc/cp/pt.c:16262
0x10219b8b tsubst_expr(tree_node*, tree_node*, int, tree_node*, bool)
	/sw/summit/gcc/6.4.0/src/gcc-6.4.0/gcc/cp/pt.c:15883
0x102189ef tsubst_expr(tree_node*, tree_node*, int, tree_node*, bool)
	/sw/summit/gcc/6.4.0/src/gcc-6.4.0/gcc/cp/pt.c:15309
0x10219487 tsubst_expr(tree_node*, tree_node*, int, tree_node*, bool)
	/sw/summit/gcc/6.4.0/src/gcc-6.4.0/gcc/cp/pt.c:15185
0x1021979b tsubst_expr(tree_node*, tree_node*, int, tree_node*, bool)
	/sw/summit/gcc/6.4.0/src/gcc-6.4.0/gcc/cp/pt.c:15371
0x10219487 tsubst_expr(tree_node*, tree_node*, int, tree_node*, bool)
	/sw/summit/gcc/6.4.0/src/gcc-6.4.0/gcc/cp/pt.c:15185
0x1021979b tsubst_expr(tree_node*, tree_node*, int, tree_node*, bool)
	/sw/summit/gcc/6.4.0/src/gcc-6.4.0/gcc/cp/pt.c:15371
0x1023a6ff instantiate_decl(tree_node*, int, bool)
	/sw/summit/gcc/6.4.0/src/gcc-6.4.0/gcc/cp/pt.c:22108
0x1023b313 instantiate_class_template_1
	/sw/summit/gcc/6.4.0/src/gcc-6.4.0/gcc/cp/pt.c:10311
0x1023b313 instantiate_class_template(tree_node*)
	/sw/summit/gcc/6.4.0/src/gcc-6.4.0/gcc/cp/pt.c:10381
0x102beacf complete_type(tree_node*)
	/sw/summit/gcc/6.4.0/src/gcc-6.4.0/gcc/cp/typeck.c:131
0x1021fb83 tsubst_copy_and_build(tree_node*, tree_node*, int, tree_node*, bool, bool)
	/sw/summit/gcc/6.4.0/src/gcc-6.4.0/gcc/cp/pt.c:17216
0x102203f7 tsubst_copy_and_build(tree_node*, tree_node*, int, tree_node*, bool, bool)
	/sw/summit/gcc/6.4.0/src/gcc-6.4.0/gcc/cp/pt.c:16564
0x10219b8b tsubst_expr(tree_node*, tree_node*, int, tree_node*, bool)
	/sw/summit/gcc/6.4.0/src/gcc-6.4.0/gcc/cp/pt.c:15883
Please submit a full bug report,
with preprocessed source if appropriate.
Please include the complete backtrace with any bug report.
See <http://gcc.gnu.org/bugs.html> for instructions.
make[2]: *** [core/unit_test/CMakeFiles/NeighborList_test_CUDA.dir/CUDA/tstNeighborList_CUDA.cpp.o] Error 1
make[1]: *** [core/unit_test/CMakeFiles/NeighborList_test_CUDA.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

Note this is in hierarchical parallel code added by @streeve in #183 although I approved via review. This code has been working now for a few months but this is the first time I've done a Cabana build on Summit since the Kokkos 3.0 release. @dalg24 has anyone reported similar issues on Summit?

--- a/core/src/Cabana_Parallel.hpp
+++ b/core/src/Cabana_Parallel.hpp
@@ -255,15 +255,15 @@ inline void neighbor_parallel_for(
 
     Kokkos::parallel_for(
         str, exec_policy, KOKKOS_LAMBDA( const index_type i ) {
-            const int nn =
+            const index_type nn =
                 NeighborList<NeighborListType>::numNeighbor( list, i );
 
-            for ( int n = 0; n < nn; ++n )
+            for ( index_type n = 0; n < nn; ++n )
             {
                 const index_type j =
                     NeighborList<NeighborListType>::getNeighbor( list, i, n );
 
-                for ( int a = n + 1; a < nn; ++a )
+                for ( index_type a = n + 1; a < nn; ++a )
                 {
                     const index_type k =
                         NeighborList<NeighborListType>::getNeighbor( list, i,
@@ -407,7 +407,7 @@ inline void neighbor_parallel_for(
         KOKKOS_LAMBDA( const typename kokkos_policy::member_type &team ) {
             index_type i = team.league_rank() + range_begin;
 
-            const int nn =
+            const index_type nn =
                 NeighborList<NeighborListType>::numNeighbor( list, i );
             Kokkos::parallel_for(
                 Kokkos::TeamThreadRange( team, nn ), [&]( const index_type n ) {
@@ -415,7 +415,7 @@ inline void neighbor_parallel_for(
                         NeighborList<NeighborListType>::getNeighbor( list, i,
                                                                      n );
 
-                    for ( int a = n + 1; a < nn; ++a )
+                    for ( index_type a = n + 1; a < nn; ++a )
                     {
                         const index_type k =
                             NeighborList<NeighborListType>::getNeighbor( list,
@@ -480,17 +480,17 @@ inline void neighbor_parallel_for(
         KOKKOS_LAMBDA( const typename kokkos_policy::member_type &team ) {
             index_type i = team.league_rank() + range_begin;
 
-            const int nn =
+            const index_type nn =
                 NeighborList<NeighborListType>::numNeighbor( list, i );
             Kokkos::parallel_for(
-                Kokkos::TeamThreadRange( team, nn ), [&]( const int n ) {
+                Kokkos::TeamThreadRange( team, nn ), [&]( const index_type n ) {
                     const index_type j =
                         NeighborList<NeighborListType>::getNeighbor( list, i,
                                                                      n );
 
                     Kokkos::parallel_for(
                         Kokkos::ThreadVectorRange( team, n + 1, nn ),
-                        [&]( const int a ) {
+                        [&]( const index_type a ) {
                             const index_type k =
                                 NeighborList<NeighborListType>::getNeighbor(
                                     list, i, a );

